From 8b3b582f86c1b8a4e21cd0afc97767c18227fe4b Mon Sep 17 00:00:00 2001
From: Nicolas Fella <nicolas.fella@gmx.de>
Date: Sat, 1 May 2021 17:25:14 +0200
Subject: [PATCH] Hide advanced output configuration when using PipeWire

PipeWire does not have support for the relevant PulseAudio modules.

See https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/771

module-switch-on-connect is less relevant since PipeWire has this behavior by default.

BUG: 433246
---
 src/modulemanager.cpp | 8 ++++++++
 src/server.cpp        | 6 ++++++
 src/server.h          | 3 +++
 3 files changed, 17 insertions(+)

diff --git a/src/modulemanager.cpp b/src/modulemanager.cpp
index acb3d92..36b22ec 100644
--- a/src/modulemanager.cpp
+++ b/src/modulemanager.cpp
@@ -7,6 +7,7 @@
 #include "modulemanager.h"
 #include "module.h"
 #include "../config.h"
+#include "server.h"
 
 #if USE_GSETTINGS
 #include "gsettingsitem.h"
@@ -100,6 +101,13 @@ ModuleManager::~ModuleManager(){};
 
 bool ModuleManager::settingsSupported() const
 {
+    // PipeWire does not (yet) have support for module-switch-on-connect and module-combine-sink
+    // Also switching streams is the default there
+    // TODO Check whether there is a PipeWire-specific way to do these
+    if (Context::instance()->server()->isPipeWire()) {
+        return false;
+    }
+
 #if USE_GCONF || USE_GSETTINGS
     return true;
 #else
diff --git a/src/server.cpp b/src/server.cpp
index e5f0345..788826d 100644
--- a/src/server.cpp
+++ b/src/server.cpp
@@ -64,6 +64,7 @@ void Server::update(const pa_server_info *info)
 {
     m_defaultSinkName = QString::fromUtf8(info->default_sink_name);
     m_defaultSourceName = QString::fromUtf8(info->default_source_name);
+    m_isPipeWire = QString::fromUtf8(info->server_name).contains("PipeWire");
 
     updateDefaultDevices();
 }
@@ -105,4 +106,9 @@ void Server::updateDefaultDevices()
     }
 }
 
+bool Server::isPipeWire() const
+{
+    return m_isPipeWire;
+}
+
 } // QPulseAudio
diff --git a/src/server.h b/src/server.h
index 8f98f25..60a8f4d 100644
--- a/src/server.h
+++ b/src/server.h
@@ -32,6 +32,8 @@ public:
     void reset();
     void update(const pa_server_info *info);
 
+    bool isPipeWire() const;
+
 Q_SIGNALS:
     void defaultSinkChanged(Sink *sink);
     void defaultSourceChanged(Source *source);
@@ -43,6 +45,7 @@ private:
     QString m_defaultSourceName;
     Sink *m_defaultSink;
     Source *m_defaultSource;
+    bool m_isPipeWire;
 };
 
 } // QPulseAudio
-- 
GitLab

